### PR TITLE
Add partition statistics

### PR DESCRIPTION
**Pull request by yuzhu**

### CODE REVIEW

**Comment by yuzhu** [comment]
Right now `partitionInfo` also contains `partitionStatistics`, we might want to move it out of partitionInfo in the future so that we can return them separately for more efficient rpcs

**Comment by yuzhu** [comment]
@gpang Will move statistics information to an object outside of the layout data so we can return them independently to the client as we discussed.

File: experimental/table/server/master/src/main/java/alluxio/master/catalog/AlluxioCatalog.java [hunk 1]
```diff
@@ -159,6 +163,39 @@ private Database getDatabaseByName(String dbName) throws NotFoundException {
...
+        return false;
+      }
+    }).map(p -> {
+      try {
+        PartitionInfo info = ProtoUtils.extractHiveLayout(p.toProto());
+        return new Pair<>(info.getPartitionName(), info.getColStats()
+            .getStatisticsList().stream().filter(e -> colNames.contains(e.getColName()))
+            .collect(Collectors.toList()));
+      } catch (InvalidProtocolBufferException e) {
+        return new Pair<>("", new ArrayList<ColumnStatisticsInfo>());
```
Comment (ZacBlanco) [inline experimental/table/server/master/src/main/java/alluxio/master/catalog/AlluxioCatalog.java]: does this denote empty/no statistics? Should it even be used by the caller?

Could we do a final pass on the result of this map, and then filter out entries with `("", ...)`?

File: experimental/table/server/master/src/main/java/alluxio/master/catalog/CatalogMaster.java [hunk 2]
```diff
@@ -101,4 +103,17 @@ boolean attachDatabase(String dbName, String dbType, CatalogConfiguration config
...
       throws IOException;
+
+  /**
+   * Get statistics on the partitions.
+   *
+   * @param dbName database name
+   * @param tableName table name
+   * @param partNamesList partition names
+   * @param colNamesList column names
+   * @return a map mapping partitions to a list of column statistics info
```
Comment (ZacBlanco) [inline experimental/table/server/master/src/main/java/alluxio/master/catalog/CatalogMaster.java]: the key is a string - is it the partition name?

**Comment by yuzhu** [comment]
@gpang separate stats from the rest of the partitionInfo. PTAL

File: experimental/table/server/master/src/main/java/alluxio/master/catalog/AlluxioCatalog.java (hunk 1 above)
Comment (yuzhu) [reply experimental/table/server/master/src/main/java/alluxio/master/catalog/AlluxioCatalog.java]: updated code no longer returns the empty entries.

File: experimental/table/server/master/src/main/java/alluxio/master/catalog/CatalogMaster.java (hunk 2 above)
Comment (yuzhu) [reply experimental/table/server/master/src/main/java/alluxio/master/catalog/CatalogMaster.java]: yes, updated javadoc

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
@ZacBlanco PTAL, thanks

File: core/transport/src/grpc/catalog/catalog_master.proto [hunk 3]
```diff
@@ -220,7 +220,7 @@ message PartitionInfo {
   optional string tableName = 3;
   optional Storage storage = 4; // storage descriptor
   map<string, string> parameters = 5;
-  optional ColumnStatisticsList col_stats = 6; // column statistics for partition
+  // optional ColumnStatisticsList col_stats = 6; // column statistics for partition
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/transport/src/grpc/catalog/catalog_master.proto]: done

**Comment by yuzhu** [comment]
@ZacBlanco Updated, PTAL

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Pull request merged**